docs/pages/application-access/guides/vnet.mdx

```diff
@@ -0,0 +1,190 @@
+version: v1
+metadata:
+  name: vnet-config
+spec:
+  custom_dns_zones:
+  - suffix: <Var name="suffix" />
+```
+
+Create the VNet config:
+
+```code
+$ tctl create vnet-config.yaml
+vnet_config has been created
+```
+
+<Admonition type="note" title="Relationship between suffix and public_addr">
+`suffix` doesn't have to point to a domain that's exactly one level above the `public_addr` of an
+app. Any level of nesting works. E.g., you can have an app under `tcp-app.foo.bar.qux.test` and the
+suffix can be set to `bar.qux.test`.
+</Admonition>
+
+## Step 2/4. Set DNS TXT record
+
+Establish the name of your cluster:
+
+```code
+$ tctl status | awk '/Cluster/ {print $2}'
+teleport.example.com
+```
+
+Set a DNS TXT record `teleport-cluster=<cluster name>` on the domain provided as `suffix` in the
+VNet config. This is required by VNet to prevent a rogue cluster admin from hijacking DNS resolution
+of arbitrary domains for unsuspecting users.
+
+Wait a couple of minutes and verify that the DNS record changes have been propagated.
+
+```code
+$ host -t TXT <Var name="suffix" />
+company.test descriptive text "teleport-cluster=teleport.example.com"
+```
+
+<Admonition type="note" title="DNS TXT and CNAME records">
+If the domain you chose for the suffix has the CNAME record set, you won't be able to add a TXT
+record. A CNAME record must be the only record for the given name. In that scenario, you can set the
+suffix to a domain one level above if possible (e.g., `company.test` instead of
+`internal.company.test`) or switch from CNAME to A records.
+</Admonition>
+
+## Step 3/4. Set `public_addr` of the app
+
+Set `public_addr` field of the application in the Application Service configuration file
+`/etc/teleport.yaml` and restart the `teleport` daemon.
+
+```yaml
+version: v3
+# …
+app_service:
+  # …
+  apps:
+  - name: "tcp-app"
+    uri: tcp://localhost:5432
+    public_addr: "<Var name="public_addr" />"
+```
+
+## Step 4/4. Connect
+
+Once you [start VNet](../../connect-your-client/vnet.mdx), you should be able to connect to the
+application over the custom `public_addr` using the application client you would normally use to
+connect to it. You might need to restart VNet if it was already running while you were making
+changes to the cluster.
+
+```code
+$ psql postgres://postgres@tcp-app.company.test/postgres
+```
+
+## Next steps
+
+### Configuring IPv4 CIDR range
+
+VNet uses `100.64.0.0/10` as the IPv4 CIDR range by default. Each resource is assigned an IP address
+from that range just before the first connection is established. If this range conflicts with other
+software deployed on devices of your users, it can be changed by setting the `ipv4_cidr_range` field
+of the VNet config.
+
+When starting, VNet fetches configs from all available clusters and arbitrarily picks a range
+provided by one of the clusters. This means that if you have a root cluster with multiple leaf
+clusters, you should update the range on each cluster.
```

Comment:
@nklaassen I feel like this is kinda subpar. What if VNet picked an arbitrary non-default range if present? Then instead of setting the range on each leaf cluster, an admin could set it only on the root cluster.

Reply:
I don't believe this paragraph is entirely accurate. Each app should be assigned an address from the `ipv4_cidr_range` of the cluster it's actually in

You might be referring to some code where we assign an IPv4 address to the TUN device from one range arbitrarily. That's true and maybe it's subpar, but it doesn't affect the IPs actually assigned to apps, that's just the address assigned to the interface that will be the source address

So e.g. if you have
- `root.example.com` with a custom range `192.168.128.0/17`
- `leaf.example.com` with the default range `100.64.0.0/10`

you might get
- TUN interface gets assigned `100.64.0.1`
- `app.root.example.com` will get `192.168.3.5`
- `app.leaf.example.com` will get `100.64.0.5`

If you dial `app.root.example.com` the source_addr will be `100.64.0.1` and the dest_addr will be `192.168.3.5`.

You're right that's probably subpar if you set the custom range on the root cluster to avoid conflicts, and then the leaf cluster, where maybe you're not even using vnet, overrides the range used to pick an IP for the TUN and stuff breaks. I'd be onboard if you want to update it to try to prefer the root cluster or any non-default range

Reply:
I added an issue for changing the logic of how the address of the TUN device is picked. Based on what you said and how the code works at the moment, it shouldn't be a big issue, but I wanted it documented somewhere. https://github.com/gravitational/teleport/issues/43343

The way the code is written at the moment means that VNet will always pick the range used by the first root cluster returned from `api/profile.ListProfileNames`. As the majority of users are logged in only to a single root cluster, this shouldn't be an issue most of the time.

---

And thanks for correcting me on that. Since in every other case the IPv4 range is taken from a config of a particular cluster, it should simplify the docs as well.